cranelift/docs/isle.md

```diff
@@ -0,0 +1,1306 @@
+  equal to the already-captured type for `x` in the type environment.
+  
+* A conjunction `(and PAT1 PAT2 ...)` checks that each subpattern is
+  compatible with the expected type.
+  
+* Integer constants are compatible with any primitive expected
+  type. (This may change in the future if we add non-numeric
+  primitives, such as strings.)
+  
+If we are able to typecheck the pattern, we have a type environment
+that is a map from variable bindings to types: e.g., `{ x: MyType, y:
+MyType2, z: u32 }`. We then typecheck the rewrite expression.
+
+* Every expression also has an expected type, from the surrounding
+  context. We check that the provided expression matches this type.
+  
+* The top-level rewrite expression must have the same type as the
+  top-level constructor in the pattern. (In other words, a term can
+  only be rewritten to another term of the same type.)
+  
+* Constructors check their return values against the expected type,
+  and typecheck their argument expressions against their parameter
+  types.
+
+* A `let` expression provides types for additional variable bindings;
+  these are added to the type environment while typechecking the
+  body. The expected type for the body is the same as the expected
+  type for the `let` itself.
+  
+### A Note on Heterogeneous Types
+
+We should illuminate one particular aspect of the ISLE type system
+that we described above. We have said that a term must be rewritten to
+another term of the same type. Note that this does *not* mean that,
+for example, a set of ISLE rules cannot be used to translate a term of
+type `T1` to a term of type `T2`. The trick is to define a top-level
+"driver" that wraps the `T1`, such that reducing this term results in
+a `T2`. Concretely:
+
+```lisp
+    (type T1 ...)
+    (type T2 ...)
+    
+    (decl Translate (T1) T2)
+    
+    (rule (Translate (T1.A ...)
+                     (T2.X ...)))
+    (rule (Translate (T1.B ...)
+                     (T2.Y ...)))
+```
+
+This gets to the heart of rewrite-system-based computation, and has
+relevance for applications of ISLE to compiler backends. A common
+technique in rewrite systems is to "kick off" a computation by
+wrapping a term in some intermediate term that then drives a series of
+reductions. Here we are using `Translate` as this top-level term. A
+difference between ISLE and some other rewrite-based instruction
+selectors is that rewrites are always driven by term reduction from
+such a toplevel term, rather than a series of equivalences directly
+from IR instruction to machine instruction forms.
+
+In other words, a conventional isel pattern engine might let one
+specify `(Inst.A ...) -> (Inst.X ...)`. Here the instruction/opcode
+type on the LHS and RHS must be the same single instruction type
+(otherwise rewrites could not be chained), and rewrite relation (which
+we wrote as `->`) is in essence a single privileged relation. One can
```

Comment:
Maybe just me, but I don't follow the sentence starting with "Here...". Does here mean in the previous generic instruction selection example, or in ISEL?

Reply:
Clarified, thanks.